src/instance.rs

```diff
@@ -32,21 +32,142 @@ pub unsafe trait PyNativeType: Sized {
     }
 }
 
-/// A Python object of known type T.
+/// A GIL-independent reference to an object allocated on the Python heap.
 ///
-/// Accessing this object is thread-safe, since any access to its API requires a `Python<'py>` GIL
-/// token. There are a few different ways to use the Python object contained:
-///  - [`Py::as_ref`](#method.as_ref) to borrow a GIL-bound reference to the contained object.
-///  - [`Py::borrow`](#method.borrow), [`Py::try_borrow`](#method.try_borrow),
-///    [`Py::borrow_mut`](#method.borrow_mut), or [`Py::try_borrow_mut`](#method.try_borrow_mut),
-///    to directly access a `#[pyclass]` value (which has RefCell-like behavior, see
-///    [the `PyCell` guide entry](https://pyo3.rs/main/class.html#pycell-and-interior-mutability)
-///    ).
-///  - Use methods directly on `Py`, such as [`Py::call`](#method.call) and
-///    [`Py::call_method`](#method.call_method).
+/// This type does not auto-dereference to the inner object because you must prove you hold the GIL to access it.
+/// Instead, call one of its methods to access the inner object:
+///  - [`Py::as_ref`], to borrow a GIL-bound reference to the contained object.
+///  - [`Py::borrow`], [`Py::try_borrow`], [`Py::borrow_mut`], or [`Py::try_borrow_mut`],
+/// to get a (mutable) reference to a contained pyclass, using a scheme similar to std's [`RefCell`].
+/// See the [`PyCell` guide entry](https://pyo3.rs/main/class.html#pycell-and-interior-mutability)
+/// for more information.
+///  - You can call methods directly on `Py` with [`Py::call`], [`Py::call_method`] and friends.
+/// These require passing in the [`Python<'py>`](crate::Python) token but are otherwise similar to the corresponding
+/// methods on [`PyAny`].
 ///
-/// See [the guide](https://pyo3.rs/main/types.html) for an explanation
-/// of the different Python object types.
+/// # Example: Storing Python objects in structs
+///
+/// As all the native Python objects only appear as references, storing them in structs doesn't work well.
+/// For example, there is no way to make something like this compile!
```

Comment:
I feel like it'd be helpful to the reader to explain what "something like this" is. I tried to word it as "no way to store `&PyDict` directly in a `#[pyclass]`.

Reply:
I think doing away with that and just "For example, this won't compile" would be better.